evm_arithmetization/tests/two_to_one_block.rs

```diff
@@ -0,0 +1,363 @@
+        &dummy_proof0,
+        dummy_pv0,
+    )?;
+
+    all_circuits.verify_aggregation(&agg_proof0)?;
+
+    // Test retrieved public values from the proof public inputs.
+    let retrieved_public_values0 = PublicValues::from_public_inputs(&agg_proof0.public_inputs);
+    assert_eq!(retrieved_public_values0, pv0);
+    assert_eq!(
+        pv0.trie_roots_before.state_root,
+        pv0.extra_block_data.checkpoint_state_trie_root
+    );
+
+    let (block_proof0, block_public_values) = all_circuits.prove_block(
+        None, // We don't specify a previous proof, considering block 1 as the new checkpoint.
+        &agg_proof0,
+        pv0.clone(),
+    )?;
+
+    let pv_block = PublicValues::from_public_inputs(&block_proof0.public_inputs);
+    assert_eq!(block_public_values, pv_block);
+
+    Ok(block_proof0)
+}
+
+#[ignore]
+#[test]
+fn test_two_to_one_block_aggregation() -> anyhow::Result<()> {
+    init_logger();
+    let some_timestamps = [127, 42, 65, 43];
+
+    let all_stark = AllStark::<F, D>::default();
+    let config = StarkConfig::standard_fast_config();
+    let all_circuits = AllRecursiveCircuits::<F, C, D>::new(
+        &all_stark,
+        &[16..17, 9..15, 12..18, 14..15, 9..10, 12..13, 17..20],
+        &config,
+    );
+
+    let unrelated_block_proofs = some_timestamps
+        .iter()
+        .map(|&ts| get_test_block_proof(ts, &all_circuits, &all_stark, &config))
+        .collect::<anyhow::Result<Vec<ProofWithPublicInputs<F, C, D>>>>()?;
+
+    unrelated_block_proofs
+        .iter()
+        .try_for_each(|bp| all_circuits.verify_block(bp))?;
+
+    let bp = unrelated_block_proofs;
+
+    {
+        // Aggregate the same proof twice
+        let aggproof_42_42 = all_circuits.prove_two_to_one_block(&bp[0], false, &bp[0], false)?;
+        all_circuits.verify_two_to_one_block(&aggproof_42_42)?;
+    }
+
+    {
+        // Binary tree reduction
+        //
+        //  A    B    C    D    Blockproofs (base case)
+        //   \  /      \  /
+        //  (A, B)    (C, D)    Two-to-one block aggregation proofs
+        //     \       /
+        //   ((A,B), (C,D))     Two-to-one block aggregation proofs
+
+        let aggproof01 = all_circuits.prove_two_to_one_block(&bp[0], false, &bp[1], false)?;
+        all_circuits.verify_two_to_one_block(&aggproof01)?;
+
+        let aggproof23 = all_circuits.prove_two_to_one_block(&bp[2], false, &bp[3], false)?;
+        all_circuits.verify_two_to_one_block(&aggproof23)?;
+
+        let aggproof0123 =
+            all_circuits.prove_two_to_one_block(&aggproof01, true, &aggproof23, true)?;
+        all_circuits.verify_two_to_one_block(&aggproof0123)?;
+
+        {
+            // Compute Merkle root from public inputs of block proofs.
+            let hash_no_pad =
+                <PoseidonGoldilocksConfig as GenericConfig<D>>::InnerHasher::hash_no_pad;
+            let two_to_one =
+                <PoseidonGoldilocksConfig as GenericConfig<D>>::InnerHasher::two_to_one;
```

Comment:
minor: may be more concise
```suggestion
            let hash_no_pad = PoseidonHash::hash_no_pad;
            let two_to_one = PoseidonHash::two_to_one;
```

Reply:
Used `PoseidonHash::hash_no_pad` and `PoseidonHash::two_to_one` directly.